astroquery/cadc/cadctap/core.py

```diff
@@ -0,0 +1,409 @@
+                raise AttributeError('Already logged in with certificate')
+                return
+            self._TapPlus__user = user
+            self._TapPlus__pwd = password
+            self.__dologin(cookie_prefix=cookie_prefix,
+                           login_url=login_url, verbose=verbose)
+        else:
+            if password is None:
+                raise AttributeError("No password")
+                return
+
+    def __dologin(self, cookie_prefix=None, login_url=None, verbose=False):
+        """
+        Notes
+        -----------------
+        Decode and set the cookie with the right key
+        """
+        self._TapPlus__isLoggedIn = False
+        response = self.__execLogin(self._TapPlus__user,
+                                    self._TapPlus__pwd,
+                                    login_url,
+                                    verbose)
+        # check response
+        connHandler = self._TapPlus__getconnhandler()
+        isError = connHandler.check_launch_response_status(response,
+                                                           verbose,
+                                                           200)
+        if isError:
+            if verbose:
+                print("Login error: " + str(response.reason))
+            raise requests.exceptions.HTTPError(
+                "Login error: " + str(response.reason))
+        else:
+            # extract cookie
+            cookie = response.read()
+            c = cookie.decode()
+            cookie = '{}="{}"'.format(cookie_prefix, c)
+            if cookie is not None:
+                self._TapPlus__isLoggedIn = True
+                connHandler.set_cookie(cookie)
+
+    def logout(self, verbose=False):
+        """Performs a logout
+        Parameters
+        ----------
+        verbose : bool, optional, default 'False'
+            flag to display information about the process
+
+        Notes
+        -----------------
+        Clear the certificate as well
+        """
+        self._TapPlus__getconnhandler()._TapConn__connectionHandler. \
+            _ConnectionHandlerCadc__certificate = None
+        if self._TapPlus__getconnhandler().cookies_set():
+            self._TapPlus__getconnhandler().unset_cookie()
+        self.__certificate = None
+        self._TapPlus__isLoggedIn = False
+
+    def __execLogin(self, usr, pwd, login_url=None, verbose=False):
+        """
+        Notes
+        -----------------
+        Send to the right login url, send not using the TapConn object
+        because it redirects instead of giving you a cookie
+        """
+        args = {
+            "username": str(usr),
+            "password": str(pwd)}
+        data = urlencode(args)
+        if login_url is None:
+            raise RuntimeError("No login URL")
+        protocol, host, port, server_context, \
+            tap_context = self._Tap__parseUrl(login_url)
+        connHandler = httplib.HTTPConnection(host, 80)

```

Comment:
this is almost certainly going to cause problems.  `astroquery` is strictly based on `requests`; direct use of `httplib` is not permitted.

Reply:
I too prefer requests but everything that is under /cadctap is supposed to be temporary until the tap library is updated at which point we want to get rid of it. We've just forked the main library, made changes so that our sub-package work and didn't waste any effort with the other things. Following your review, I've made some easy, obvious changes but I believe that moving away from httplib requires more effort that it's worth. I would rather make those changes to the main tap library in this package but we were told that someone else is already doing that.
Comments?

Reply:
@andamian - this is a topic where I think having a discussion with the multiple stakeholders would be very useful, definitely beyond the scope of this PR.

Reply:
Right, thanks for the reminder.  Indeed, we can let this pass and try to get it fixed elsewhere.